extensions/eclipse-che-theia-plugin-remote/src/node/plugin-remote-node-impl.ts

```diff
@@ -0,0 +1,253 @@
+        // tslint:disable-next-line: no-any
+        const plugin = (this.pluginManager as any).registry.get(pluginId);
+        if (plugin) {
+            plugin.rawModel = pluginPackage;
+        }
+    }
+
+    // tslint:disable-next-line: no-any
+    deasyncPromise(promise: Promise<any>) {
+        // tslint:disable-next-line: no-any
+        let result: any;
+        // tslint:disable-next-line: no-any
+        let error: any;
+        let done = false;
+        promise.then(res => {
+            result = res;
+        }, err => {
+            error = err;
+        }).then(() => {
+            done = true;
+        });
+        deasync.loopWhile(() => !done);
+        if (error) {
+            throw error;
+        }
+        return result;
+    }
+
+    async $definePluginExports(hostId: string, pluginId: string, proxyNames: string[]): Promise<void> {
+        // tslint:disable-next-line: no-any
+        const pluginManagerInternal = (this.pluginManager as any);
+
+        // add into the activatedPlugins stuff
+        // tslint:disable-next-line: no-any
+        const activatedPlugin: any = {};
+        // tslint:disable-next-line: no-any
+        const proxyExports: any = {};
+        activatedPlugin.exports = proxyExports;
+        const remoteBrowser = this.pluginRemoteBrowser;
+        const deasyncPromise = this.deasyncPromise;
+        // add proxy
+        const callId = this.callId++;
+        proxyNames.forEach(entryName => {
+            console.log(`Proxyfing exports ${hostId} ${pluginId}/${entryName}`);
+            // tslint:disable-next-line: no-any
+            proxyExports[entryName] = (...args: any[]) => {
+                // need to keep arguments only if arguments have functions as functions can't be propagated remotely
+                const hasFunctions = args.some(arg => typeof arg === 'function');
+                if (hasFunctions) {
+                    const functions = new Map<number, Function>();
+                    args.forEach((arg, index) => {
+                        const type = typeof arg;
+                        if (type === 'function') {
+                            functions.set(index, arg);
+                        }
+                    });
+                    // keep function arguments
+                    this.calls.set(callId, { functions });
+                }
+
+                // remote call for this method
+                return deasyncPromise(remoteBrowser.$callMethod(hostId, pluginId, callId, entryName, ...args));
+            };
+        });
+        pluginManagerInternal.activatedPlugins.set(pluginId, activatedPlugin);
+        await pluginManagerInternal.activatePlugin(pluginId);
+    }
+
+    // tslint:disable-next-line: no-any
+    async $callLocalMethod(callId: number, index: number, ...args: any[]): Promise<any> {
+        const callInfo = this.calls.get(callId);
+        if (callInfo) {
+            const invocatioObject = callInfo.functions.get(index);
```

Comment:
I think "functionToInvoke" would be a better name. The object lifetime is longer than a single invocation.